ndk/src/font.rs

```diff
@@ -0,0 +1,401 @@
+        if 0 < value && value <= 1000 {
+            Ok(Self(value))
+        } else {
+            Err(TryFromU16Error(()))
+        }
+    }
+
+    pub const fn value(&self) -> u16 {
+        self.0
+    }
+
+    pub const THIN: FontWeight = FontWeight(ffi::AFONT_WEIGHT_THIN as u16);
+    pub const EXTRA_LIGHT: FontWeight = FontWeight(ffi::AFONT_WEIGHT_EXTRA_LIGHT as u16);
+    pub const LIGHT: FontWeight = FontWeight(ffi::AFONT_WEIGHT_LIGHT as u16);
+    pub const NORMAL: FontWeight = FontWeight(ffi::AFONT_WEIGHT_NORMAL as u16);
+    pub const MEDIUM: FontWeight = FontWeight(ffi::AFONT_WEIGHT_MEDIUM as u16);
+    pub const SEMI_BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_SEMI_BOLD as u16);
+    pub const BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_BOLD as u16);
+    pub const EXTRA_BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_EXTRA_BOLD as u16);
+    pub const BLACK: FontWeight = FontWeight(ffi::AFONT_WEIGHT_BLACK as u16);
+    pub const MAX: FontWeight = FontWeight(ffi::AFONT_WEIGHT_MAX as u16);
+}
+
+impl fmt::Display for FontWeight {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            FontWeight::THIN => writeln!(f, "Thin"),
+            FontWeight::EXTRA_LIGHT => writeln!(f, "Extra Light (Ultra Light)"),
+            FontWeight::LIGHT => writeln!(f, "Light"),
+            FontWeight::NORMAL => writeln!(f, "Normal (Regular)"),
+            FontWeight::MEDIUM => writeln!(f, "Medium"),
+            FontWeight::SEMI_BOLD => writeln!(f, "Semi Bold (Demi Bold)"),
+            FontWeight::BOLD => writeln!(f, "Bold"),
+            FontWeight::EXTRA_BOLD => writeln!(f, "Extra Bold (Ultra Bold)"),
+            FontWeight::BLACK => writeln!(f, "Black (Heavy)"),
+            _ => writeln!(f, "{}", self.0),
+        }
+    }
+}
+
+/// The error type returned when an invalie font weight value is passed.
+#[derive(Debug)]
+pub struct TryFromU16Error(());
+
+impl fmt::Display for TryFromU16Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            fmt,
+            "font weight must be positive and less than or equal to 1000"
+        )
+    }
+}
+
+impl std::error::Error for TryFromU16Error {}
+
+impl TryFrom<u16> for FontWeight {
+    type Error = TryFromU16Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        FontWeight::new(value)
+    }
+}
+
+/// A 4-byte integer representing an OpenType axis tag.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct AxisTag(u32);
+
+impl AxisTag {
+    pub const fn from_be(value: u32) -> std::result::Result<Self, TryFromU32Error> {
+        // Each byte in a tag must be in the range 0x20 to 0x7E.
+        // See https://learn.microsoft.com/en-us/typography/opentype/spec/otff#data-types for details.
+        macro_rules! check_in_valid_range {
+            ($($byte: expr),+) => {
+                $(
+                    if !(0x20 <= ($byte) && ($byte) <= 0x7E) {
```

Comment:
Hmm `char::is_ascii()` only checks the upper-bound, not the lower one.

Reply:
Fixed! Also, I've just found that `0x20 ' '` is only for trailing padding, so I added checks for that as well.
```rust
        // Each byte in a tag must be in the range 0x20 to 0x7E. A space character cannot be
        // followed by a non-space character. A tag must have one to four non-space characters.
        // See https://learn.microsoft.com/en-us/typography/opentype/spec/otff#data-types for details.

        let bytes = value.to_be_bytes();

        macro_rules! check_if_valid {
            ($e:expr ; $($f:expr)+) => {
                if !(bytes[$e] as char).is_ascii_graphic() {
                    return if true $(&& bytes[$f] == b' ')+ {
                        Ok(Self(value))
                    } else {
                        Err(TryFromU32Error(()))
                    };
                }
            };
        }

        // A tag must have one to four non-space characters.
        if !(bytes[0] as char).is_ascii_graphic() {
            return Err(TryFromU32Error(()));
        }

        check_if_valid!(1; 1 2 3);
        check_if_valid!(2; 2 3);
        check_if_valid!(3; 3);

        Ok(Self(value))
```